ghcide/src/Development/IDE/Core/Shake.hs

```diff
@@ -183,8 +183,8 @@ data ShakeExtras = ShakeExtras
     -- positions in a version of that document to positions in the latest version
     -- First mapping is delta from previous version and second one is an
     -- accumlation of all previous mappings.
-    ,inProgress :: Var (HMap.HashMap NormalizedFilePath Int)
-    -- ^ How many rules are running for each file
+    ,inProgress :: forall a . NormalizedFilePath -> Action a -> Action a
+    -- ^ Report progress for a rule
     ,progressUpdate :: ProgressEvent -> IO ()
```

Comment:
Did you consider just putting a whole ProgressReporting in ShakeExtras, so you can keep it a bit more abstract?